# Predicting State-wise EV Adoption  
## Notebook 02: Data Cleaning

### Purpose
This notebook focuses on the preprocessing stage of the dataset, where we handle missing values, outliers, data format, etc. required in order to make the dataset fit for the following stages.

The goal is to make a clean dataset which can be processed without issues in the upcoming stages of the project by creating a seperate dataset from the original dataset.

In [1]:
# get a view of the original dataset
import pandas as pd
df = pd.read_csv("../Data/ev_vs_petrol_dataset_v3.csv")
df.head()

,country,region,year,vehicle_segment,powertrain_type,ev_sales,petrol_car_sales,diesel_car_sales,total_vehicle_sales,ev_market_share,...,avg_ev_range_km,fuel_price_usd_per_liter,electricity_price_usd_per_kwh,gdp_per_capita,urban_population_percent,co2_emissions_transport_mt,ev_subsidy_usd,emission_regulation_score,ev_growth_rate_yoy,is_ev_dominant
0,Australia,Oceania,2010,commercial,ICE,5,92877,61921,154803,0.00,...,124,1.09,0.149,51977,88.8,88.7,0,30.4,0.00,0
1,Australia,Oceania,2010,mass_market,ICE,57,535933,73089,609079,0.01,...,124,1.09,0.149,51977,88.8,88.7,0,30.4,0.00,0
2,Australia,Oceania,2010,premium,ICE,37,235282,20462,255781,0.01,...,124,1.09,0.149,51977,88.8,88.7,0,30.4,0.00,0
3,Australia,Oceania,2011,commercial,ICE,11,98092,65395,163498,0.01,...,133,1.09,0.163,52807,88.9,88.3,0,30.8,120.00,0
4,Australia,Oceania,2011,mass_market,ICE,129,569679,77684,647492,0.02,...,133,1.09,0.163,52807,88.9,88.3,0,30.8,126.32,0


In [2]:
# creates a new copy of the dataset to prevent accidental overwriting
df_clean = df.copy()

In [3]:
# understand the dataset statistics and features
df_clean.info()
df_clean.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 22 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   country                        1200 non-null   object 
 1   region                         1200 non-null   object 
 2   year                           1200 non-null   int64  
 3   vehicle_segment                1200 non-null   object 
 4   powertrain_type                1200 non-null   object 
 5   ev_sales                       1200 non-null   int64  
 6   petrol_car_sales               1200 non-null   int64  
 7   diesel_car_sales               1200 non-null   int64  
 8   total_vehicle_sales            1200 non-null   int64  
 9   ev_market_share                1200 non-null   float64
 10  charging_stations              1200 non-null   int64  
 11  fast_chargers_share            1200 non-null   float64
 12  avg_ev_range_km                1200 non-null   i

,year,ev_sales,petrol_car_sales,diesel_car_sales,total_vehicle_sales,ev_market_share,charging_stations,fast_chargers_share,avg_ev_range_km,fuel_price_usd_per_liter,electricity_price_usd_per_kwh,gdp_per_capita,urban_population_percent,co2_emissions_transport_mt,ev_subsidy_usd,emission_regulation_score,ev_growth_rate_yoy,is_ev_dominant
count,1200.000000,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1200.000000,1.200000e+03,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000
mean,2017.500000,6.360236e+04,7.733063e+05,1.409369e+05,9.778455e+05,6.327992,5.693898e+04,14.437750,265.872500,1.342239,0.169508,35651.525000,74.609000,190.837000,2910.940000,58.761860,64.002958,0.018333
std,4.611694,4.079506e+05,1.756934e+06,2.913472e+05,2.211799e+06,13.231723,3.359383e+05,10.057529,121.994051,0.363461,0.079298,23899.013679,14.858811,363.100639,2477.306742,20.092871,71.386797,0.134210
min,2010.000000,5.000000e+00,1.727000e+03,1.500000e+02,1.748000e+04,0.000000,0.000000e+00,0.000000,106.000000,0.482900,0.066990,1359.000000,30.900000,5.400000,0.000000,19.484000,-40.960000,0.000000
25%,2013.750000,2.002500e+02,8.870500e+04,1.951075e+04,1.300465e+05,0.077500,4.917500e+02,6.000000,150.000000,1.097000,0.098750,11077.250000,64.175000,30.175000,0.000000,43.600000,16.372500,0.000000
50%,2017.500000,2.731500e+03,2.294785e+05,5.136100e+04,2.905130e+05,0.830000,4.090000e+03,12.900000,238.500000,1.352000,0.158000,37329.000000,79.900000,94.250000,2794.500000,59.350000,44.165000,0.000000
75%,2021.250000,1.990500e+04,6.385320e+05,1.111910e+05,7.624482e+05,5.832500,2.251800e+04,20.950000,380.250000,1.568500,0.220250,50930.000000,84.075000,159.575000,4973.750000,74.825000,80.907500,0.000000
max,2025.000000,7.670056e+06,1.477369e+07,2.014595e+06,1.884998e+07,95.000000,4.338106e+06,49.300000,507.000000,2.153000,0.379040,103437.000000,98.500000,1900.800000,8952.000000,95.602000,300.000000,1.000000


In [4]:
# Rename columns for clarity
df_clean.rename(columns={
    "total_vehicle_sales": "total_vehicle_sales(units)",
    "ev_sales": "ev_sales(units)",
    "petrol_car_sales": "petrol_car_sales(units)",
    "diesel_car_sales": "diesel_car_sales(units)",
    "ev_market_share": "ev_market_share(%)",
    "charging_stations": "charging_stations(units)",
    "fast_chargers_share": "fast_chargers_share(%)",
    "avg_ev_range_km": "average_ev_range(km)",
    "fuel_price_usd_per_liter": "fuel_price(USD/ltr)",
    "electricity_price_usd_per_kwh": "electricity_price(USD/kwh)",
    "gdp_per_capita": "gdp(USD/person)",
    "urban_population_percent": "urban_population(%)",
    "co2_emissions_transport_mt": "co2_emissions(MT)",
    "ev_subsidy_usd": "ev_subsidy(USD)",
    "emission_regulation_score": "emission_regulation(scale_0-100)",
    "ev_growth_rate_yoy": "ev_growth_rate_yoy(%)",
    "is_ev_dominant": "ev_domination(0/1)"
}, inplace=True)


# Remove unnecessary columns
df_clean.drop(columns=[
    "powertrain_type"
], inplace=True)


# Confirm changes
df_clean.columns

Index(['country', 'region', 'year', 'vehicle_segment', 'ev_sales(units)',
       'petrol_car_sales(units)', 'diesel_car_sales(units)',
       'total_vehicle_sales(units)', 'ev_market_share(%)',
       'charging_stations(units)', 'fast_chargers_share(%)',
       'average_ev_range(km)', 'fuel_price(USD/ltr)',
       'electricity_price(USD/kwh)', 'gdp(USD/person)', 'urban_population(%)',
       'co2_emissions(MT)', 'ev_subsidy(USD)',
       'emission_regulation(scale_0-100)', 'ev_growth_rate_yoy(%)',
       'ev_domination(0/1)'],
      dtype='object')

In [5]:
# standardise column names
df_clean.columns = df_clean.columns.str.lower().str.strip().str.replace(" ", "_")

In [6]:
# checks for duplicates of the dataset
print(df_clean.duplicated().sum())

# drop duplicates if any exist securely
initial_len = len(df_clean)
df_clean.drop_duplicates(inplace=True)
print(f"Duplicates removed: {initial_len - len(df_clean)}")


0
Duplicates removed: 0


In [7]:
# checking for missing values(no missing values)
df_clean.isnull().sum()

country                             0
region                              0
year                                0
vehicle_segment                     0
ev_sales(units)                     0
petrol_car_sales(units)             0
diesel_car_sales(units)             0
total_vehicle_sales(units)          0
ev_market_share(%)                  0
charging_stations(units)            0
fast_chargers_share(%)              0
average_ev_range(km)                0
fuel_price(usd/ltr)                 0
electricity_price(usd/kwh)          0
gdp(usd/person)                     0
urban_population(%)                 0
co2_emissions(mt)                   0
ev_subsidy(usd)                     0
emission_regulation(scale_0-100)    0
ev_growth_rate_yoy(%)               0
ev_domination(0/1)                  0
dtype: int64

In [8]:
# data type correction for categorical columns
df_clean["country"] = df_clean["country"].astype("category")
df_clean["region"] = df_clean["region"].astype("category")
df_clean["vehicle_segment"] = df_clean["vehicle_segment"].astype("category")

# data type correction for integer columns
df_clean["year"] = df_clean["year"].astype(int)
df_clean["total_vehicle_sales(units)"] = df_clean["total_vehicle_sales(units)"].astype(int)
df_clean["ev_sales(units)"] = df_clean["ev_sales(units)"].astype(int)
df_clean["petrol_car_sales(units)"] = df_clean["petrol_car_sales(units)"].astype(int)
df_clean["diesel_car_sales(units)"] = df_clean["diesel_car_sales(units)"].astype(int)
df_clean["charging_stations(units)"] = df_clean["charging_stations(units)"].astype(int)
df_clean["average_ev_range(km)"] = df_clean["average_ev_range(km)"].astype(int)
df_clean["gdp(usd/person)"] = df_clean["gdp(usd/person)"].astype(int)
df_clean["ev_subsidy(usd)"] = df_clean["ev_subsidy(usd)"].astype(int)

# data type correction for float columns
df_clean["ev_market_share(%)"] = df_clean["ev_market_share(%)"].astype(float)
df_clean["ev_growth_rate_yoy(%)"] = df_clean["ev_growth_rate_yoy(%)"].astype(float)
df_clean["electricity_price(usd/kwh)"] = df_clean["electricity_price(usd/kwh)"].astype(float)
df_clean["fuel_price(usd/ltr)"] = df_clean["fuel_price(usd/ltr)"].astype(float)
df_clean["co2_emissions(mt)"] = df_clean["co2_emissions(mt)"].astype(float)
df_clean["emission_regulation(scale_0-100)"] = df_clean["emission_regulation(scale_0-100)"].astype(float)
df_clean["urban_population(%)"] = df_clean["urban_population(%)"].astype(float)
df_clean["fast_chargers_share(%)"] = df_clean["fast_chargers_share(%)"].astype(float)

# data type correction for binary column
df_clean["ev_domination(0/1)"] = df_clean["ev_domination(0/1)"].astype(int)

In [9]:
# logical validations for all integer and float columns

print("EV sales negative:",
      (df_clean["ev_sales(units)"] < 0).sum())

print("Petrol sales negative:",
      (df_clean["petrol_car_sales(units)"] < 0).sum())

print("Diesel sales negative:",
      (df_clean["diesel_car_sales(units)"] < 0).sum())

print("Total sales <= 0:",
      (df_clean["total_vehicle_sales(units)"] <= 0).sum())

print("Market share outside 0–100:",
      ((df_clean["ev_market_share(%)"] < 0) |
       (df_clean["ev_market_share(%)"] > 100)).sum())

print("Charging Stations:",
      (df_clean["charging_stations(units)"] < 0).sum())

print("Fast charger share outside 0–100:",
      ((df_clean["fast_chargers_share(%)"] < 0) |
       (df_clean["fast_chargers_share(%)"] > 100)).sum())

print("Fuel price <= 0:",
      (df_clean["fuel_price(usd/ltr)"] <= 0).sum())

print("Electricity price <= 0:",
      (df_clean["electricity_price(usd/kwh)"] <= 0).sum())

print("Average EV Range <= 0:",
      (df_clean["average_ev_range(km)"] <= 0).sum())

print("GDP <= 0:",
      (df_clean["gdp(usd/person)"] <= 0).sum())

print("Urban population outside 0–100:",
      ((df_clean["urban_population(%)"] < 0) |
       (df_clean["urban_population(%)"] > 100)).sum())

print("CO2 emissions <= 0:",
      (df_clean["co2_emissions(mt)"] <= 0).sum())

print("Subsidy < 0:",
      (df_clean["ev_subsidy(usd)"] < 0).sum())

print("Regulation score outside 0–100:",
      ((df_clean["emission_regulation(scale_0-100)"] < 0) |
       (df_clean["emission_regulation(scale_0-100)"] > 100)).sum())


EV sales negative: 0
Petrol sales negative: 0
Diesel sales negative: 0
Total sales <= 0: 0
Market share outside 0–100: 0
Charging Stations: 0
Fast charger share outside 0–100: 0
Fuel price <= 0: 0
Electricity price <= 0: 0
Average EV Range <= 0: 0
GDP <= 0: 0
Urban population outside 0–100: 0
CO2 emissions <= 0: 0
Subsidy < 0: 0
Regulation score outside 0–100: 0


In [10]:
# filter dataset to preserve only valid, logically correct records

print(f"Row count before filtering bad constraints: {len(df_clean)}")

df_clean = df_clean[
    (df_clean["ev_sales(units)"] >= 0) &
    (df_clean["petrol_car_sales(units)"] >= 0) &
    (df_clean["diesel_car_sales(units)"] >= 0) &
    (df_clean["total_vehicle_sales(units)"] > 0) &
    (df_clean["ev_market_share(%)"].between(0, 100)) &
    (df_clean["fast_chargers_share(%)"].between(0, 100)) &
    (df_clean["fuel_price(usd/ltr)"] > 0) &
    (df_clean["electricity_price(usd/kwh)"] > 0) &
    (df_clean["average_ev_range(km)"] > 0) &
    (df_clean["gdp(usd/person)"] > 0) &
    (df_clean["urban_population(%)"].between(0, 100)) &
    (df_clean["co2_emissions(mt)"] > 0) &
    (df_clean["ev_subsidy(usd)"] >= 0) &
    (df_clean["emission_regulation(scale_0-100)"].between(0, 100))
].copy()

print(f"Row count after filtering bad constraints: {len(df_clean)}")


Row count before filtering bad constraints: 1200
Row count after filtering bad constraints: 1200


In [11]:
# sort dataset in alphabetical and ascending order
df_clean.sort_values(by=["country", "year"], inplace=True)
df_clean.reset_index(drop=True, inplace=True)

In [12]:
# detect outliers using IQR method
numeric_columns = [
    "ev_sales(units)",
    "total_vehicle_sales(units)",
    "petrol_car_sales(units)",
    "diesel_car_sales(units)",
    "charging_stations(units)",
    "fast_chargers_share(%)",
    "average_ev_range(km)",
    "fuel_price(usd/ltr)",
    "electricity_price(usd/kwh)",
    "gdp(usd/person)",
    "urban_population(%)",
    "co2_emissions(mt)",
    "ev_subsidy(usd)",
    "emission_regulation(scale_0-100)",
    "ev_growth_rate_yoy(%)",
    "ev_market_share(%)"
]

for col in numeric_columns:  
    Q1 = df_clean[col].quantile(0.25)
    Q3 = df_clean[col].quantile(0.75)
    
    IQR = Q3 - Q1
    
    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR
    
    outliers = df_clean[
        (df_clean[col] < lower) |
        (df_clean[col] > upper)
    ]
    
    print(col, "outliers:", len(outliers))

ev_sales(units) outliers: 165
total_vehicle_sales(units) outliers: 145
petrol_car_sales(units) outliers: 133
diesel_car_sales(units) outliers: 146
charging_stations(units) outliers: 117
fast_chargers_share(%) outliers: 12
average_ev_range(km) outliers: 0
fuel_price(usd/ltr) outliers: 0
electricity_price(usd/kwh) outliers: 0
gdp(usd/person) outliers: 0
urban_population(%) outliers: 27
co2_emissions(mt) outliers: 96
ev_subsidy(usd) outliers: 0
emission_regulation(scale_0-100) outliers: 0
ev_growth_rate_yoy(%) outliers: 97
ev_market_share(%) outliers: 165


Observation:
There are 10 numeric columns with outliers. With the highest outlier count being 165 for `ev_market_share(%)` as well as `ev_sales(units)`, and the lowest outlier count being 12 from `fast_chargers_share(%)`. Other outlier columns include `urban_population(%)`, `co2_emissions(mt) outlier`, `charging_stations(units)`, `petrol_car_sales(units)`, `diesel_car_sales(units)`, `ev_growth_rate_yoy(%)`, and `total_vehicle_sales(units)`. 

In [13]:
# cross-check the final clean dataset
df_clean.info()
df_clean.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 21 columns):
 #   Column                            Non-Null Count  Dtype   
---  ------                            --------------  -----   
 0   country                           1200 non-null   category
 1   region                            1200 non-null   category
 2   year                              1200 non-null   int32   
 3   vehicle_segment                   1200 non-null   category
 4   ev_sales(units)                   1200 non-null   int32   
 5   petrol_car_sales(units)           1200 non-null   int32   
 6   diesel_car_sales(units)           1200 non-null   int32   
 7   total_vehicle_sales(units)        1200 non-null   int32   
 8   ev_market_share(%)                1200 non-null   float64 
 9   charging_stations(units)          1200 non-null   int32   
 10  fast_chargers_share(%)            1200 non-null   float64 
 11  average_ev_range(km)              1200 non-null   int32 

,year,ev_sales(units),petrol_car_sales(units),diesel_car_sales(units),total_vehicle_sales(units),ev_market_share(%),charging_stations(units),fast_chargers_share(%),average_ev_range(km),fuel_price(usd/ltr),electricity_price(usd/kwh),gdp(usd/person),urban_population(%),co2_emissions(mt),ev_subsidy(usd),emission_regulation(scale_0-100),ev_growth_rate_yoy(%),ev_domination(0/1)
count,1200.000000,1.200000e+03,1.200000e+03,1.200000e+03,1.200000e+03,1200.000000,1.200000e+03,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000
mean,2017.500000,6.360236e+04,7.733063e+05,1.409369e+05,9.778455e+05,6.327992,5.693898e+04,14.437750,265.872500,1.342239,0.169508,35651.525000,74.609000,190.837000,2910.940000,58.761860,64.002958,0.018333
std,4.611694,4.079506e+05,1.756934e+06,2.913472e+05,2.211799e+06,13.231723,3.359383e+05,10.057529,121.994051,0.363461,0.079298,23899.013679,14.858811,363.100639,2477.306742,20.092871,71.386797,0.134210
min,2010.000000,5.000000e+00,1.727000e+03,1.500000e+02,1.748000e+04,0.000000,0.000000e+00,0.000000,106.000000,0.482900,0.066990,1359.000000,30.900000,5.400000,0.000000,19.484000,-40.960000,0.000000
25%,2013.750000,2.002500e+02,8.870500e+04,1.951075e+04,1.300465e+05,0.077500,4.917500e+02,6.000000,150.000000,1.097000,0.098750,11077.250000,64.175000,30.175000,0.000000,43.600000,16.372500,0.000000
50%,2017.500000,2.731500e+03,2.294785e+05,5.136100e+04,2.905130e+05,0.830000,4.090000e+03,12.900000,238.500000,1.352000,0.158000,37329.000000,79.900000,94.250000,2794.500000,59.350000,44.165000,0.000000
75%,2021.250000,1.990500e+04,6.385320e+05,1.111910e+05,7.624482e+05,5.832500,2.251800e+04,20.950000,380.250000,1.568500,0.220250,50930.000000,84.075000,159.575000,4973.750000,74.825000,80.907500,0.000000
max,2025.000000,7.670056e+06,1.477369e+07,2.014595e+06,1.884998e+07,95.000000,4.338106e+06,49.300000,507.000000,2.153000,0.379040,103437.000000,98.500000,1900.800000,8952.000000,95.602000,300.000000,1.000000


In [14]:
# view of the final dataset
df_clean.head()

,country,region,year,vehicle_segment,ev_sales(units),petrol_car_sales(units),diesel_car_sales(units),total_vehicle_sales(units),ev_market_share(%),charging_stations(units),...,average_ev_range(km),fuel_price(usd/ltr),electricity_price(usd/kwh),gdp(usd/person),urban_population(%),co2_emissions(mt),ev_subsidy(usd),emission_regulation(scale_0-100),ev_growth_rate_yoy(%),ev_domination(0/1)
0,Australia,Oceania,2010,commercial,5,92877,61921,154803,0.00,0,...,124,1.09,0.149,51977,88.8,88.7,0,30.4,0.00,0
1,Australia,Oceania,2010,mass_market,57,535933,73089,609079,0.01,0,...,124,1.09,0.149,51977,88.8,88.7,0,30.4,0.00,0
2,Australia,Oceania,2010,premium,37,235282,20462,255781,0.01,0,...,124,1.09,0.149,51977,88.8,88.7,0,30.4,0.00,0
3,Australia,Oceania,2011,commercial,11,98092,65395,163498,0.01,105,...,133,1.09,0.163,52807,88.9,88.3,0,30.8,120.00,0
4,Australia,Oceania,2011,mass_market,129,569679,77684,647492,0.02,105,...,133,1.09,0.163,52807,88.9,88.3,0,30.8,126.32,0


In [15]:
# save cleaned dataset
df_clean.to_csv(
    "../Data/ev_adoption_cleaned.csv",
    index=False
)

### Observations Summary

- Missing values verified
- Renamed columns
- Redundant data removed(`powertrain_type`)
- Data types standardized
- Logical constraints validated
- Outliers detected and preserved
- Data leakage features and multicollinearity features will be removed during the prediction phase, since they are needed for the Feature Engineering and EDA phase
- Note: We will also keep in mind of the features generated from Feature Engineering phase that might cause data leakage and multicollinearity as well
- Duplicate records non existent
- Clean dataset created for further processing

### Next Step

Proceed to Notebook 03: Feature Engineering